engineio/asyncio_server.py

```diff
@@ -220,24 +231,37 @@ def start_background_task(self, target, *args, **kwargs):
         """
         return asyncio.ensure_future(target(*args, **kwargs))
 
-    async def sleep(self, seconds=0):
+    async def sleep(self, seconds=0, until_shutdown=False):
         """Sleep for the requested amount of time using the appropriate async
         model.
 
         This is a utility function that applications can use to put a task to
         sleep without having to worry about using the correct call for the
         selected async mode.
 
+        :param seconds: the number of seconds to sleep
+        :param until_shutdown: interrupt sleep if server is shutdowned
+
         Note: this method is a coroutine.
         """
-        return await asyncio.sleep(seconds)
+        if not until_shutdown:
+            return await asyncio.sleep(seconds)
+
+        # Sleep a short time and recheck if server is still running.
+        fractional, integer = math.modf(seconds * 10)
+        for i in range(int(integer)):
+            await asyncio.sleep(0.1)
+            if not self.running:
+                return
+        return await asyncio.sleep(fractional)
```

Comment:
What you are doing here is a somewhat busy wait loop. This is going to add CPU load, so it's not something that I find acceptable. I would actually prefer to leave the sleep function working as a sleep function, and build a better mechanism for the task service to be awaken if it needs to exit, maybe with an `Event` object or similar.